Add Data Prepper Overview (#1453)
opensearch-project/documentation-website: _clients/data-prepper/index.md

Data Prepper lets users build custom pipelines to improve the operational view of applications. Two common uses for Data Prepper are trace and log analytics. [Trace analytics]({{site.url}}{{site.baseurl}}/observability-plugin/trace/index/) can help you visualize the flow of events and identify performance problems, and [log analytics]({{site.url}}{{site.baseurl}}/observability-plugin/log-analytics/) can improve searching, analyzing and provide insights into your application.

## Concepts

Data Prepper is compromised of **Pipelines** that collect and filter data based on the components set within the pipeline. Each component is pluggable, enabling you to use your own custom implementation of each component. These components include: 

- One [source](#source)
- One or more[sinks](#sink)
- (Optional) One [buffer](#buffer)
- (Optional) One or more[processors](#processor)

A single instance of Data Prepper can have one or more pipelines. 

Each pipeline definition contains two required components **source** and **sink**. If buffers and processors are missing from the Data Prepper pipeline, Data Prepper uses the default buffer and a no-op processor. 

### Source 

> Source is the input component of a pipeline that defines the mechanism through which a Data Prepper pipeline will consume events. A pipeline can have only one source. The source can consume events either by receiving the events over HTTP or HTTPS or reading from external endpoints like OTeL Collector for traces and metrics and S3. Source have their own configuration options based on the format of the events (such as string, json,  cloudwatch logs, or open telemetry trace). The source component consumes events and writes them to the buffer component. 

Reviewer: ```suggestion
Source is the input component of a pipeline that defines the mechanism through which a Data Prepper pipeline will consume events. A pipeline can have only one source. The source can consume events either by receiving the events over HTTP or HTTPS or by reading from external endpoints like OTeL Collector for traces and metrics and Amazon Simple Storage Service (Amazon S3). Sources have their own configuration options based on the format of the events (such as string, JSON, Amazon CloudWatch logs, or open telemetry trace). The source component consumes events and writes them to the buffer component. 
```

> Source is the input component of a pipeline that defines the mechanism through which a Data Prepper pipeline will consume events. A pipeline can have only one source. The source can consume events either by receiving the events over HTTP or HTTPS or reading from external endpoints like OTeL Collector for traces and metrics and S3. Source have their own configuration options based on the format of the events (such as string, json,  cloudwatch logs, or open telemetry trace). The source component consumes events and writes them to the buffer component. 

Reviewer: Should the first sentence start with "The source component"? In the third sentence, should it read something like "...or by reading from external endpoints, like OTeL Collector for traces, metrics, and Amazon Simple Storage Service (Amazon S3)"?

### Buffer

> The buffer component acts as the layer between the source and the sink. Buffer can be either in-memory or disk-based. The default buffer uses an in-memory queue bounded by the number of events, called `bounded_blocking`. If the buffer component is not explicitly mentioned in the pipeline configuration, Data Prepper uses the default `bounded_blocking`.

Reviewer: Second sentence: "The buffer..."?

> The buffer component acts as the layer between the source and the sink. Buffer can be either in-memory or disk-based. The default buffer uses an in-memory queue bounded by the number of events, called `bounded_blocking`. If the buffer component is not explicitly mentioned in the pipeline configuration, Data Prepper uses the default `bounded_blocking`.

Reviewer: ```suggestion
The buffer component acts as the layer between the source and the sink. Buffer can be either in-memory or disk based. The default buffer uses an in-memory queue called `bounded_blocking` that is bounded by the number of events. If the buffer component is not explicitly mentioned in the pipeline configuration, Data Prepper uses the default `bounded_blocking`.
```

### Sink

> Sink is the output component of a pipeline that defines the destination(s) to which a Data Prepper pipeline publishes events. A sink destination could be services such as OpenSearch, S3, or another Data Prepper pipeline. When using another Data Prepper pipeline as the sink, you can chain multiple pipelines together based on the needs to the data. Sink contains it's own configurations options based on the destination type.

Reviewer: Should the first sentence start with "The sink component..."? Same in the third sentence: "When using another Data Prepper pipeline as the sink component..."? And in the fourth sentence: "The sink component contains..."? Confirm that my changes to the second sentence retain the intended meaning.

Author: @natebower: For Source and Sink we need to indicate that they are input and output components, in order for technical users to understand the parameters they are configuring for each.

Which of these do you prefer:

A. "The sink output component defines the destination..."
B. "Sink is the output component that defines the destination(s) to which a Data Prepper pipeline publishes events."

> Sink is the output component of a pipeline that defines the destination(s) to which a Data Prepper pipeline publishes events. A sink destination could be services such as OpenSearch, S3, or another Data Prepper pipeline. When using another Data Prepper pipeline as the sink, you can chain multiple pipelines together based on the needs to the data. Sink contains it's own configurations options based on the destination type.

Reviewer: ```suggestion
Sink is the output component of a pipeline that defines the destination(s) to which a Data Prepper pipeline publishes events. A sink destination could be a service, such as OpenSearch or Amazon S3, or another Data Prepper pipeline. When using another Data Prepper pipeline as the sink, you can chain multiple pipelines together based on the needs of the data. Sink contains its own configuration options based on the destination type.
```

### Processor

> Processors are units within the Data Prepper pipeline that can filter, transform, and enrich events into your desired format before publishing the record to the sink. The a processor is not defined in the pipeline configuration, the events publish in the format defined in the source component. You can have more than on processor within a pipeline. When using multiple processors, the processors are executed in the order they are defined inside the pipeline spec.

Reviewer: ```suggestion
Processors are units within the Data Prepper pipeline that can filter, transform, and enrich events into your desired format before publishing the record to the sink component. The processor is not defined in the pipeline configuration; the events publish in the format defined in the source component. You can have more than one processor within a pipeline. When using multiple processors, the processors are run in the order they are defined inside the pipeline spec.
```

> Processors are units within the Data Prepper pipeline that can filter, transform, and enrich events into your desired format before publishing the record to the sink. The a processor is not defined in the pipeline configuration, the events publish in the format defined in the source component. You can have more than on processor within a pipeline. When using multiple processors, the processors are executed in the order they are defined inside the pipeline spec.

Reviewer: In the first sentence, it looks like "into" does not agree with all three verbs. Would "using" work? Last word: Can "spec" be "specification"?

Author: Works for me.

> ## Sample Pipeline configurations

Reviewer: Should "Pipeline" be lowercase?

> ## Sample Pipeline configurations

Author: ```suggestion
## Sample pipeline configurations
```

To understand how all pipeline components function within a Data Prepper configuration, see the following examples. Each pipeline configuration uses a `yaml` file format.

### Minimal component

> This pipeline configuration reads from file source and writes to that same source. It uses the default options for buffer and processor.

Reviewer: Should this be "This pipeline configuration reads from the file source and writes to that same source. It uses the default options for the buffer and processor"?

> This pipeline configuration reads from file source and writes to that same source. It uses the default options for buffer and processor.

Author: ```suggestion
This pipeline configuration reads from the file source and writes to that same source. It uses the default options for the buffer and processor.
```

```yml
sample-pipeline:
  source:
    file:
        path: <path/to/input-file>
  sink:
    - file:
        path: <path/to/output-file>
```

### All components

> The following pipeline uses a source that reads string events from the `input-file`. The source then pushes the data to buffer bounded by max size of `1024`. The pipeline configured to have `4` workers each of them reading maximum of `256` events from the buffer for every `100 milliseconds`. Each worker executes the `string_converter` processor and write the output of the processor to the `output-file`.

Reviewer: ```suggestion
The following pipeline uses a source that reads string events from the `input-file`. The source then pushes the data to the buffer, bounded by a max size of `1024`. The pipeline is configured to have `4` workers, each of them reading a maximum of `256` events from the buffer for every `100 milliseconds`. Each worker runs the `string_converter` processor and writes the output of the processor to the `output-file`.
```
